crates/sui-types/src/passkey_authenticator.rs

```diff
@@ -0,0 +1,191 @@
+use fastcrypto::{error::FastCryptoError, traits::ToFromBytes};
+use once_cell::sync::OnceCell;
+use passkey::types::webauthn::CollectedClientData;
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use shared_crypto::intent::IntentMessage;
+use std::hash::Hash;
+use std::hash::Hasher;
+use std::sync::Arc;
+
+#[cfg(test)]
+#[path = "unit_tests/passkey_authenticator_test.rs"]
+mod passkey_authenticator_test;
+
+/// An zk login authenticator with all the necessary fields.
+#[derive(Debug, Clone, JsonSchema, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct PasskeyAuthenticator {
+    authenticator_data: Vec<u8>, // should size limit be enforced?
+    client_data: Vec<u8>,        // should size limit be enforced?
+    user_signature: Signature,
+    #[serde(skip)]
+    pub bytes: OnceCell<Vec<u8>>,
+}
+
+impl PasskeyAuthenticator {
+    // test only
+    pub fn new(
+        authenticator_data: Vec<u8>,
+        client_data: Vec<u8>,
+        user_signature: Signature,
+    ) -> Self {
+        Self {
+            authenticator_data,
+            client_data,
+            user_signature,
+            bytes: OnceCell::new(),
+        }
+    }
+
+    pub fn get_pk(&self) -> SuiResult<PublicKey> {
+        PublicKey::try_from_bytes(
+            SignatureScheme::Secp256r1,
+            self.user_signature.public_key_bytes(),
+        )
+        .map_err(|_| SuiError::InvalidAuthenticator)
+    }
+}
+
+/// Necessary trait for [struct SenderSignedData].
+impl PartialEq for PasskeyAuthenticator {
+    fn eq(&self, other: &Self) -> bool {
+        self.as_ref() == other.as_ref()
+    }
+}
+
+/// Necessary trait for [struct SenderSignedData].
+impl Eq for PasskeyAuthenticator {}
+
+/// Necessary trait for [struct SenderSignedData].
+impl Hash for PasskeyAuthenticator {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.as_ref().hash(state);
+    }
+}
+
+impl AuthenticatorTrait for PasskeyAuthenticator {
+    fn verify_user_authenticator_epoch(
+        &self,
+        _epoch: EpochId,
+        _max_epoch_upper_bound_delta: Option<u64>,
+    ) -> SuiResult {
+        Ok(())
+    }
+
+    /// Verify an intent message of a transaction with an zk login authenticator.
+    fn verify_claims<T>(
+        &self,
+        intent_msg: &IntentMessage<T>,
+        author: SuiAddress,
+        _aux_verify_data: &VerifyParams,
+        _zklogin_inputs_cache: Arc<VerifiedDigestCache<ZKLoginInputsDigest>>,
+    ) -> SuiResult
+    where
+        T: Serialize,
+    {
+        let client_data_json: CollectedClientData = serde_json::from_slice(&self.client_data)
+            .map_err(|_| SuiError::InvalidSignature {
+                error: "get client data".to_string(),
+            })?;
+        let parsed_challenge =
+            Base64UrlUnpadded::decode_vec(&client_data_json.challenge).map_err(|e| {
+                SuiError::InvalidSignature {
+                    error: e.to_string(),
+                }
+            })?;
```

Comment:
Instead of having unpacking this here, can we do that at the type level, so the authenticator already has this parsed out at deserialization time?

Reply:
good idea, we can do this at deserialization.